[clang][RISCV] Fix crash on VLS calling convention

4vtomat opened the pull request:
This patch handle struct of fixed vector and struct of array of fixed
vector correctly for VLS calling convention in EmitFunctionProlog,
EmitFunctionEpilog and EmitCall.

stack on: https://github.com/llvm/llvm-project/pull/147173

wangpc-pp commented on clang/test/CodeGen/RISCV/riscv-vector-callingconv-llvm-ir.c:
```diff
@@ -143,52 +143,82 @@ void __attribute__((riscv_vls_cc)) test_too_large(int32x64_t arg) {}
 void __attribute__((riscv_vls_cc(256))) test_st_i32x4x9_256(struct st_i32x4x9 arg) {}
+
+// CHECK-LLVM: define dso_local riscv_vls_cc(128) <vscale x 8 x i32> @test_function_prolog_epilog(<vscale x 8 x i32> %arg.coerce) #0 {
+// CHECK-LLVM: entry:
```
Use CHECK-XXX-LABEL and CHECK-XXX-NEXT?

4vtomat replied:
Right, thanks for notice that!

wangpc-pp reviewed (approved):
LGTM.

topperc commented on clang/test/CodeGen/RISCV/riscv-vector-callingconv-llvm-ir.c:
```diff
@@ -143,52 +143,82 @@ void __attribute__((riscv_vls_cc)) test_too_large(int32x64_t arg) {}
+// CHECK-LLVM-NEXT: entry:
+// CHECK-LLVM-NEXT:   %retval = alloca %struct.st_i32x4_arr4, align 16
+// CHECK-LLVM-NEXT:   %arg = alloca %struct.st_i32x4_arr4, align 16
+// CHECK-LLVM-NEXT:   store <vscale x 8 x i32> %arg.coerce, ptr %arg, align 16
```
Can you manually add a CHECK line for the declaration of struct.st_i32x4_arr4

4vtomat replied:
The type itself does not have `vscale` but the argument passed in has type `<vscale x 8 x i32>`, if we cannot use a scalable store, how do we store the input argument?

topperc replied:
Maybe you can extract the fixed vector from the scalable vector using llvm.vector.extract?

4vtomat replied:
What if it's vector tuple type? It needs n `llvm.tuple.insert`, and if the scalable vector type in the tuple type is not aligned with the one obtained by using `getContainerForFixedLengthVector`, it would be illegal.
The reason it might happen is `ABI_VLEN` might be different to `MinVLen` or `MaxELen`

4vtomat replied:
Yeah, if we have homogeneous fixed vector in a struct, it would be classified as a vector tuple

efriedma-quic commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -1329,6 +1329,19 @@ static llvm::Value *CreateCoercedLoad(Address Src, llvm::Type *Ty,
+      // clang/test/CodeGen/RISCV/riscv-vector-callingconv-llvm-ir.c
+      // for more information.
+      Src = Src.withElementType(Ty);
+      return CGF.Builder.CreateLoad(Src);
```
CreateCoercedLoad/CreateCoercedStore have two purposes:

- Try to make the type of the load/store match the type of other operations on the memory, as indicated by SrcTy.
- Ensure that we don't load/store past the end of the object.

The first is an optimization, but the second is required for correctness: accessing past the end of an object is undefined behavior.  You can't skip past the relevant checks just because there's a scalable type involved.

4vtomat committed:
[clang][RISCV] Fix crash on VLS calling convention

This patch handle struct of fixed vector and struct of array of fixed
vector correctly for VLS calling convention in EmitFunctionProlog,
EmitFunctionEpilog and EmitCall.

4vtomat commented:
Rebase

efriedma-quic reviewed (commented):
I'm not really happy with the way the code is organized for "coerced" load/store involving scalable types.

At the point where we're analyzing the calling convention; we have a bunch of information: exactly what kind of coercion we're doing, whether the sizes match, whether there are any unusual types involved.  We're throwing away that information, and then trying to re-derive it later.

So I think:

- The target code should compute whether the sizes are guaranteed to match.  If they do, "coerced" load/store simplifies to just a plain load/store, and we can skip all these complicated computations involving the sizes of the source/destination.
- If some types need target-specific code to lower arguments, that should be its own ABIArgInfo::Kind, and the target-independent code should call into a target-specific implementation in clang/lib/CodeGen/Targets/RISCV.cpp.

4vtomat commented:
> I'm not really happy with the way the code is organized for "coerced" load/store involving scalable types.
> 
> At the point where we're analyzing the calling convention; we have a bunch of information: exactly what kind of coercion we're doing, whether the sizes match, whether there are any unusual types involved. We're throwing away that information, and then trying to re-derive it later.
> 
> So I think:
> 
> * The target code should compute whether the sizes are guaranteed to match.  If they do, "coerced" load/store simplifies to just a plain load/store, and we can skip all these complicated computations involving the sizes of the source/destination.
> * If some types need target-specific code to lower arguments, that should be its own ABIArgInfo::Kind, and the target-independent code should call into a target-specific implementation in clang/lib/CodeGen/Targets/RISCV.cpp.

@efriedma-quic Thanks for your suggestion, I agree that not to mess up the code in coerced load/store with target specific logic is better for maintainability, and yeah maybe we need a new `ABIArgInfo::Kind` in this case and have target hooks for this one.

4vtomat commented:
Gentle ping~

topperc commented on clang/lib/CodeGen/ABIInfo.cpp:
```diff
@@ -244,6 +244,15 @@ ABIInfo::getOptimalVectorMemoryType(llvm::FixedVectorType *T,
   return T;
 }
 
+llvm::Value *ABIInfo::CreateCoercedLoad(Address SrcAddr, const ABIArgInfo &AI,
```
Function names should start with lower case.

4vtomat replied:
I see, somehow the static function of `CreateCoercedLoad` and `CreateCoercedStore` are upper case lol

topperc commented on clang/lib/CodeGen/Targets/RISCV.cpp:
```diff
@@ -780,6 +786,179 @@ ABIArgInfo RISCVABIInfo::extendType(QualType Ty, llvm::Type *CoerceTy) const {
+                                             CodeGenFunction &CGF) const {
+  llvm::Type *Ty = AI.getCoerceToType();
+  llvm::Type *SrcTy = Src.getElementType();
+  llvm::StructType *SrcSTy = dyn_cast<llvm::StructType>(SrcTy);
```
Use `cast` if you don't expect it to fail.

topperc commented on clang/lib/CodeGen/Targets/RISCV.cpp:
```diff
@@ -780,6 +786,179 @@ ABIArgInfo RISCVABIInfo::extendType(QualType Ty, llvm::Type *CoerceTy) const {
+                                      const ABIArgInfo &AI, bool DestIsVolatile,
+                                      CodeGenFunction &CGF) const {
+  llvm::Type *SrcTy = Val->getType();
+  llvm::StructType *DstSTy = dyn_cast<llvm::StructType>(Dst.getElementType());
```
`cast`

4vtomat committed:
fixup! [clang][RISCV] Fix crash on VLS calling convention

efriedma-quic reviewed (approved):
LGTM
